### ISSUE TITLE
Windows ARM64 clang_cl_arm64 compilation issue

### ISSUE THREAD

**Issue by bigboss-trades** [redorav/hlslpp#91]
I've been working on compiling some projects targeting Windows arm64 (Qualcomm Snapdragon) that use this library as a submodule. The projects compile when using the msvc_arm64  toolset in visual studio 2022, but they do not compile when I use the clang_cl_arm64 toolset. The error I get is originates from the include of arm64_neon.h inside neon.h. The errors are all related to typedef redefinitions that happen in arm64_neon.h. 

The include causing the problem is

https://github.com/redorav/hlslpp/blob/94f266ac7f8626019017dee9c077a3252f5f0731/include/hlsl%2B%2B/platforms/neon.h#L10-L14

It seems the clang-cl arm64 toolset already provides definitions for these and this include creates a conflict. I've tested a few things, but the most reasonable fixes I've found are to modify line 10 to be either

```
#if defined(_M_ARM64) && defined(_MSVC_TRADITIONAL)
```
or
```
#if defined(_M_ARM64) && !defined(__clang__)
```

Either of these get me past the compiler error.

**Comment by redorav** [comment]
Hi @bigboss-trades, thanks for reaching out. I'm not sure I understand your problem fully. I get that the combination of the clang toolset with arm64 is what makes the error appear but I don't understand what the error itself is. Could you give a bit more detail about your setup and the error messages?

**Comment by redorav** [comment]
Looking at it again I can see that clang and/or gcc might favor the other include arm_neon instead of arm64_neon which is a Microsoft thing. I could invert the logic so if clang or gcc are present we use the proper include, otherwise use the other MS-specific one which was already a bit special back when I first started it

**Comment by bigboss-trades** [comment]
Hi @redorav, thanks for the quick reply. My setup is a Surface Laptop 7, which is running Windows 11 arm64 on a Snapdragon X plus processor. I have the arm version of Visual Studio 2022 installed with the C++ Desktop Development with C++ option, along with "C++ CMake Tools for Windows" and "C++ Clang Compiler for Windows"

I created the simplest program I could to demonstrate the error. Below you'll see my IDE with 
1. A basic CMakeLists.txt file that includes the hlslpp include directory
2. A simple main.cpp file that prints hello world and includes hlsl++.h
3. The CMakeSettings.json

When I try to build with the config below, specifically the **clang_cl_arm64** toolset (inheritEnvrionments) I get the "undefined symbol" errors for neon_dupqrf64, neon_dupqf32, and neon_dupqr32 from lld-link shown in the lower left pane. When I made the modifications to neon.h I mentioned in my initial report, it builds.

If I change the toolset to **msvc_arm64** then the code builds with no modifications to neon.h needed.

I don't know enough about arm_neon.h and arm64_neon.h to tell exactly what's happening here... I'm thinking that arm64_neon.h does not have all the symbols needed by clang, but if you include both arm_neon.h and arm64_neon.h when using clang you get a conflict. 

I hope this answers more questions than it raises... your patience is really appreciated. I don't do a lot on github or in c++ to be honest so please bear with me as I try to get this info to you as best I can.

<img width="1151" alt="image" src="https://github.com/user-attachments/assets/f98bd4ea-edaf-455b-a925-96cd4697fe0d" />

CMakeSettings.json:
``` 
{
  "configurations": [
    {
      "name": "arm64-clang-Debug",
      "generator": "Ninja",
      "configurationType": "Debug",
      "buildRoot": "${projectDir}\\out\\build\\${name}",
      "installRoot": "${projectDir}\\out\\install\\${name}",
      "cmakeCommandArgs": "",
      "buildCommandArgs": "",
      "ctestCommandArgs": "",
      "inheritEnvironments": [ "clang_cl_arm64" ]
    }
  ]
}
```

**Comment by bigboss-trades** [comment]
Hi @redorav, thanks for the quick reply. My setup is a Surface Laptop 7, which is running Windows 11 arm64 on a Snapdragon X plus processor. I have the arm version of Visual Studio 2022 installed with the C++ Desktop Development with C++ option, along with "C++ CMake Tools for Windows" and "C++ Clang Compiler for Windows"

I created the simplest program I could to demonstrate the error. Below you'll see my IDE with 
1. A basic CMakeLists.txt file that includes the hlslpp include directory
2. A simple main.cpp file that prints hello world and includes hlsl++.h
3. The CMakeSettings.json

When I try to build with the config below, specifically the **clang_cl_arm64** toolset (inheritEnvrionments) I get the "undefined symbol" errors for neon_dupqrf64, neon_dupqf32, and neon_dupqr32 from lld-link shown in the lower left pane. When I made the modifications to neon.h I mentioned in my initial report, it builds.

If I change the toolset to **msvc_arm64** then the code builds with no modifications to neon.h needed.

I don't know enough about arm_neon.h and arm64_neon.h to tell exactly what's happening here... I'm thinking that arm64_neon.h does not have all the symbols needed by clang, but if you include both arm_neon.h and arm64_neon.h when using clang you get a conflict. 

I hope this answers more questions than it raises... your patience is really appreciated. I don't do a lot on github or in c++ to be honest so please bear with me as I try to get this info to you as best I can.

<img width="1151" alt="image" src="https://github.com/user-attachments/assets/faf56e71-d0b7-4329-9956-01cd133534b9" />

CMakeSettings.json:
``` 
{
  "configurations": [
    {
      "name": "arm64-clang-Debug",
      "generator": "Ninja",
      "configurationType": "Debug",
      "buildRoot": "${projectDir}\\out\\build\\${name}",
      "installRoot": "${projectDir}\\out\\install\\${name}",
      "cmakeCommandArgs": "",
      "buildCommandArgs": "",
      "ctestCommandArgs": "",
      "inheritEnvironments": [ "clang_cl_arm64" ]
    }
  ]
}
```

**Comment by redorav** [comment]
Thank you for the extra info, the screenshot tells me what I need to know. The arm64 header is MSVC only and it looks like clang cannot make sense of those function names like neon_dupqrf64. I had to remap a lot of defines to get it to work and those must me MS-only. It's a bit of a mess being honest but has an easy fix. I never tried the combo clangcl + arm64 but I can put a new config to test. I'm away right now but as soon as I'm in front of the laptop I'll put a fix in.

**Comment by bigboss-trades** [comment]
I think newer versions of the MSVC tools automatically handle including arm64_neon.h when including arm_neon.h. I took a look at arm_neon.h in **C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.42.34433\include**  and the first thing it does is check for including arm64_neon.h. (code snippet below)

I tested deleting lines 10-120 in neon.h in hlslpp, and replacing it with simply #include <arm_neon.h> and I can build the hello world project using both clang_cl_arm64 and msvc_arm64... so maybe the root cause of the original workaround has been fixed?
 
```
/***
*   arm_neon.h - declarations/definitions for ARM NEON specific intrinsics
*
*       Copyright (c) Microsoft Corporation. All rights reserved.
*
*Purpose:
*       This include file contains the declarations for ARM NEON intrinsic functions
*
****/

#pragma once

#if defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || defined (_M_ARM64EC)
#include <arm64_neon.h>
#else  /* defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || defined(_M_ARM64EC) */
```

**Comment by bigboss-trades** [comment]
I can confirm that automatic include of arm64_neon.h in arm_neon.h was added in MSVC v142. 

To confirm this, I installed the following components in Visual Studio Installer
* MSVC v143 - VS 2022 C++ ARM64 Build Tools (latest) _already installed_
* MSVC v142 - VS 2019 C++ ARM64 Build Tools (v14.29-16.11)
* MSVC v141 - VS 2017 C++ ARM64 Build Tools (v14.16)

I then compared the contents of arm_neon.h from the 3 different packages in **C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC**

The statement including arm64_neon.h was added to arm_neon.h with MSVC v142. 

<img width="1151" alt="image" src="https://github.com/user-attachments/assets/65466996-3435-4574-be37-f1c1177bd850" />

**Comment by redorav** [comment]
Hey @bigboss-trades this is a really good investigation. Based on your findings I have decided on the fix in 6f5274c66132e8f951c400103d897582b8f21491. It keeps things simpler and I think that fixes your use case. If you could update and check that everything's working correctly that'd be great, I tested in VS2022 and it was ok.

**Comment by bigboss-trades** [comment]
I just updated to the 3.6 release containing this fix, and it solves my problem both in the hello world program, and the downstream projects I was working with earlier. Thank you!
